## Importing Libraries

In [1]:
import pandas as pd
import numpy as np
import pickle
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
from sklearn import svm
from sklearn.metrics import accuracy_score
from sklearn.linear_model import LogisticRegression

## Importing Dataset
[Dataset Link - Kaggle Diabetes Dataset](https://www.kaggle.com/datasets/uciml/pima-indians-diabetes-database)

In [2]:
diabetes_dataframe = pd.read_csv('datasets/diabetes.csv')

## Data Preparation

#### Statistcial Measure of Dataset

In [3]:
diabetes_dataframe.describe()

Pregnancies     Glucose  BloodPressure  SkinThickness     Insulin  \
count   768.000000  768.000000     768.000000     768.000000  768.000000   
mean      3.845052  120.894531      69.105469      20.536458   79.799479   
std       3.369578   31.972618      19.355807      15.952218  115.244002   
min       0.000000    0.000000       0.000000       0.000000    0.000000   
25%       1.000000   99.000000      62.000000       0.000000    0.000000   
50%       3.000000  117.000000      72.000000      23.000000   30.500000   
75%       6.000000  140.250000      80.000000      32.000000  127.250000   
max      17.000000  199.000000     122.000000      99.000000  846.000000   

              BMI  DiabetesPedigreeFunction         Age     Outcome  
count  768.000000                768.000000  768.000000  768.000000  
mean    31.992578                  0.471876   33.240885    0.348958  
std      7.884160                  0.331329   11.760232    0.476951  
min      0.000000                  0.078000   21.000000    0.000000  
25%     27.300000                  0.243750   24.000000    0.000000  
50%     32.000000                  0.372500   29.000000    0.000000  
75%     36.600000                  0.626250   41.000000    1.000000  
max     67.100000                  2.420000   81.000000    1.000000

#### Details about Dataset

In [4]:
diabetes_dataframe.shape

(768, 9)

In [5]:
diabetes_dataframe.columns

Index(['Pregnancies', 'Glucose', 'BloodPressure', 'SkinThickness', 'Insulin',
       'BMI', 'DiabetesPedigreeFunction', 'Age', 'Outcome'],
      dtype='object')

"We have data for 768 patients, encompassing 9 features: 'Pregnancies', 'Glucose', 'BloodPressure', 'SkinThickness', 'Insulin', 'BMI', 'DiabetesPedigreeFunction', 'Age', and 'Outcome'."

#### Cleaning Dataset

In [6]:
diabetes_dataframe.isnull().sum()

Pregnancies                 0
Glucose                     0
BloodPressure               0
SkinThickness               0
Insulin                     0
BMI                         0
DiabetesPedigreeFunction    0
Age                         0
Outcome                     0
dtype: int64

The dataset is clean, as no missing values were detected.

#### Outcome Insights

In [7]:
diabetes_dataframe['Outcome'].value_counts()

Outcome
0    500
1    268
Name: count, dtype: int64

Out of 768 patient, 500 patient are non diabetic and 268 are diabetic

In [8]:
diabetes_dataframe.groupby('Outcome').mean()

Pregnancies     Glucose  BloodPressure  SkinThickness     Insulin  \
Outcome                                                                      
0           3.298000  109.980000      68.184000      19.664000   68.792000   
1           4.865672  141.257463      70.824627      22.164179  100.335821   

               BMI  DiabetesPedigreeFunction        Age  
Outcome                                                  
0        30.304200                  0.429734  31.190000  
1        35.142537                  0.550500  37.067164

Patients with diabetes (Outcome=1) have higher average values for features such as Glucose, Insulin, and BMI compared to patients without diabetes (Outcome=0).

In [9]:
Data = diabetes_dataframe.drop(columns=['Outcome','SkinThickness'], axis =1)
Outcome = diabetes_dataframe['Outcome']

In [10]:
Data

Pregnancies  Glucose  BloodPressure  Insulin   BMI  \
0              6      148             72        0  33.6   
1              1       85             66        0  26.6   
2              8      183             64        0  23.3   
3              1       89             66       94  28.1   
4              0      137             40      168  43.1   
..           ...      ...            ...      ...   ...   
763           10      101             76      180  32.9   
764            2      122             70        0  36.8   
765            5      121             72      112  26.2   
766            1      126             60        0  30.1   
767            1       93             70        0  30.4   

     DiabetesPedigreeFunction  Age  
0                       0.627   50  
1                       0.351   31  
2                       0.672   32  
3                       0.167   21  
4                       2.288   33  
..                        ...  ...  
763                     0.171   63  
764                     0.340   27  
765                     0.245   30  
766                     0.349   47  
767                     0.315   23  

[768 rows x 7 columns]

In [11]:
Outcome

0      1
1      0
2      1
3      0
4      1
      ..
763    0
764    0
765    0
766    1
767    0
Name: Outcome, Length: 768, dtype: int64

### Data Standardization

In [12]:
scaler = StandardScaler()
scaler.fit(Data)

Data= scaler.transform(Data)

In [13]:
Data

array([[ 0.63994726,  0.84832379,  0.14964075, ...,  0.20401277,
         0.46849198,  1.4259954 ],
       [-0.84488505, -1.12339636, -0.16054575, ..., -0.68442195,
        -0.36506078, -0.19067191],
       [ 1.23388019,  1.94372388, -0.26394125, ..., -1.10325546,
         0.60439732, -0.10558415],
       ...,
       [ 0.3429808 ,  0.00330087,  0.14964075, ..., -0.73518964,
        -0.68519336, -0.27575966],
       [-0.84488505,  0.1597866 , -0.47073225, ..., -0.24020459,
        -0.37110101,  1.17073215],
       [-0.84488505, -0.8730192 ,  0.04624525, ..., -0.20212881,
        -0.47378505, -0.87137393]])

### Data Splitting

In [14]:
Data_train, Data_test, Outcome_train , Outcome_test = train_test_split(Data, Outcome, test_size = 0.1, stratify= Outcome, random_state = 2)

### Training the SVM Model

In [15]:
svm_classifier = svm.SVC(kernel='linear')
svm_classifier.fit(Data_train, Outcome_train)

SVC(kernel='linear')

#### Model Evaluation

In [16]:
Data_train_prediction = svm_classifier.predict(Data_train)
training_data_accuracy = accuracy_score(Data_train_prediction, Outcome_train)

Data_test_prediction = svm_classifier.predict(Data_test)
test_data_accuracy = accuracy_score(Data_test_prediction, Outcome_test)

print("Training Data Accuracy is :",training_data_accuracy)
print("Testing Data Accuracy is :",test_data_accuracy)

Training Data Accuracy is : 0.784370477568741
Testing Data Accuracy is : 0.7792207792207793


#### Testing the SVM Model

In [17]:
input_data = np.array([85,66,29,0,26.6,0.351,31]).reshape(1, -1)
scaler.fit(input_data) 

std_data = scaler.transform(input_data)

prediction = svm_classifier.predict(std_data)

if (prediction[0] == 0):
    print('The person is safe')
else:
    print('The person is may get diabetic')


The person is safe


### Training Logistic Model

In [18]:
model = LogisticRegression()
model.fit(Data_train, Outcome_train)

LogisticRegression()

#### Model Evaluation

In [19]:
accuracy = model.score(Data_test, Outcome_test)
print("Testing Data Accuracy is :",accuracy)

Testing Data Accuracy is : 0.7662337662337663


### Testing the Logistic Model

In [20]:
input_data = np.array([0, 150, 72, 0, 33.6, 0.627, 18]).reshape(1, -1)

std_data = scaler.transform(input_data)

prediction = model.predict(std_data)
probability = model.predict_proba(std_data)[:, 1]

if probability < 0.2:
    print("Probability of Having Diabetes is Very Low")
elif probability < 0.4:
    print("LowProbability of Having Diabetes is Low")
elif probability < 0.6:
    print( "Probability of Having Diabetes is Moderate")
elif probability < 0.8:
    print("Probability of Having Diabetes is High")
else:
    print("Probability of Having Diabetes is Very High")


Probability of Having Diabetes is Very High


In [22]:
pickle.dump(model, open('ML Models/Diabetics_logistic_model.sav','wb'))